File: zero_compression.py
from random import randint
from tqdm import tqdm


# test examples
# inn = [0, 0, 0, 0, 0, 0, 0, 7, 5, 0, 0, 10, 0]
# out = [0, 7, 7, 5, 0, 2, 10, 0, 1]

# inn = [0, 7, 5]
# out = [0, 1, 7, 5]

# inn = [8, 9, 3, 4, 5, 6, 2, 3, 5, 6]
# out = [8, 9, 3, 4, 5, 6, 2, 3, 5, 6]

# inn = [1, 3, 0, 0, 4, 0, 0, 0]
# out = [1, 3, 0, 2, 4, 0, 3]

inn = [1, 0, 2, 0, 3, 0, 4, 0, 5, 0]
out = [1, 0, 1, 2, 0, 1, 3, 0, 1, 4, 0, 1, 5, 0, 1]


# NOTE: keves 0-nal nem tomorit, hanem hizlal
def compress(inn):
    cnt = 0

    for n in inn:
        # zero
        if n == 0:
            cnt += 1
        # non zero
        else:
            # was zero before: [... 0 N]
            if cnt == 0:
                yield n
            # was no zero before [... N]
            else:
                yield 0
                yield cnt
                yield n
                cnt = 0
    # last item was zero: [... 0]
    if cnt != 0:
        yield 0
        yield cnt
        cnt = 0


histo = {}

# --- PARAMETERS ---
howmany = range(300 * 1000)
# howmany = range(3)
# input_length = range(0, randint(8, 60))
input_length = range(60)

for _ in tqdm(howmany, ascii=True):
    inn = [randint(0, 255) for _ in input_length]

    # add some zeros
    # zero_prob = 66
    # zero_prob = randint(0, 100)
    zero_prob = randint(66, 100)
    inn = [x if randint(0, 100) > zero_prob else 0 for x in inn]

    out = list(compress(inn))

    # see results
    # print(
    #     f'in:\t{inn}',
    #     f'out:\t{out}',
    #     f'input size:\t{len(inn)}',
    #     'compression:\t{:.2f}'.format(len(out) / len(inn)),
    #     'size change:\t{0:+d}'.format(len(out) - len(inn)),
    #     f'zero prob:\t{zero_prob}'
    #     '\n',
    #     sep='\n'
    # )

    # compression stats
    try:
        # L = len(out) - len(inn) # size change
        L = '{:.3}'.format(len(out) / len(inn))  # compression ratio
        histo[L] += 1
    except KeyError:
        histo[L] = 1

print(histo)


# print(
#     list(compress(inn)),
#     list(compress(inn)) == out,
#     sep='\n'
# )

File: zero_compression.c
#include <stdio.h>
#include <stdint.h>

/** 
 * Replace (0, ..., 0) series of zero numbers to a (0, <number of zeros>) pair. Non-zero numbers
 * are unchanged. The compressed array can be decoded without data loss.
 *
 * @param [in] input: Number array to be compressed.
 * @param [in] input_length: Length of input array.
 * @param [out] compressed: Compressed version of input array. This should be 
 * around 1.5x size of the input array. In the worst case scenario, output is bigger than the input.
 * @retval Length of 'compressed' array.
 *
 * Example Usage:
 * @code
 * uint8_t arr[] = {1, 0, 0, 0, 4, 0, 0, 0};
 * uint8_t result[32];
 * uint8_t result_size;
 * result_size = compress_zeros(arr, 8, result);
 * // result == {1, 0, 3, 4, 0, 3}
 * @endcode
 */
uint8_t compress_zeros(uint8_t *input, uint8_t input_length, uint8_t *compressed)
{
    // zero counter
    uint8_t cnt = 0;

    // pointer used to write into the output array
    // NOTE: this could be removed, its only really neccessary for calculating the output array size
    uint8_t *out = compressed;

    // loop the input array
    for (uint8_t *n = input; n < (input + input_length); n++)
    {
        // NOTE: *n is the current input value
        // keep counting zeros until a non-zero item comes
        if (*n == 0x00)
        {
            cnt++;
        }
        // non-zero
        else
        {
            // was zero before: [... 0 N]
            if (cnt == 0)
            {
                // append the number as is
                *out++ = *n;
                // NOTE: the instruction above is shorthand for *out = *n; out++;
            }
            // was no zero before: [... N]
            else
            {
                // append (0, #zeros, some-number)
                *out++ = 0x00;
                *out++ = cnt;
                *out++ = *n;

                cnt = 0;
            }
        }
    }
    // last item was zero: [... 0]
    if (cnt != 0)
    {
        // append (0, #zeros)
        *out++ = 0x00;
        *out++ = cnt;
    }

    // length of the compressed array
    return (out - compressed);
}

void NOP()
{
    return;
}

// ---
// real test examples

// uint8_t arr[] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0};
// out = ?

// uint8_t arr[] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0};
// out = ?

// uint8_t arr[] = {0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 1, 1, 1};
// out = ?

// uint8_t arr[] = {0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 2, 1, 0, 1, 0};
// out = ?

// ---
// baked test examples

// uint8_t arr[] = {0, 0, 0, 0, 0, 0, 0, 7, 5, 0, 0, 10, 0};
// out = {0, 7, 7, 5, 0, 2, 10, 0, 1};

// uint8_t arr[] = {0, 7, 5};
// out = {0, 1, 7, 5};

// uint8_t arr[] = {8, 9, 3, 4, 5, 6, 2, 3, 5, 6};
// out = {8, 9, 3, 4, 5, 6, 2, 3, 5, 6};

uint8_t arr[] = {1, 0, 0, 0, 4, 0, 0, 0};
// out = {1, 0, 3, 4, 0, 3};

// uint8_t arr[] = {1, 0, 2, 0, 3, 0, 4, 0, 5, 0};
// out = {1, 0, 1, 2, 0, 1, 3, 0, 1, 4, 0, 1, 5, 0, 1};

uint8_t arr_size = sizeof(arr) / sizeof(uint8_t);

uint8_t result[32];

uint8_t result_size;

void main()
{
    // a quick heuristic to test if its worth compressing
    uint8_t count_zero_n = 0;
    uint8_t count_zeros = 0;

    for (uint8_t i = 0; i < arr_size - 1; i++)
    {
        if (arr[i] == 0)
        {
            // increment if encountering a (0, non-zero) pair,
            // check ((NOT A) AND B) truth table to verify
            count_zero_n += (!arr[i] && arr[i + 1]);

            count_zeros++;
        }
    }

    NOP();

    result_size = compress_zeros(arr, arr_size, result);

    NOP();
}